Problem with listing blobs by using of ListBlobsHierarchy()

### Bug Report

- import path of package in question `github.com/Azure/azure-sdk-for-go/sdk/storage/azblob`
- SDK version 
	github.com/Azure/azure-sdk-for-go/sdk/azidentity v0.12.0
	github.com/Azure/azure-sdk-for-go/sdk/storage/azblob v0.2.0
	github.com/Azure/azure-sdk-for-go/sdk/to v0.1.1 // indirect
	github.com/aws/aws-sdk-go v1.40.45
- Go version: go version go1.16.6 darwin/amd64
- Azurite: mcr.microsoft.com/azure-storage/azurite:3.14.3

In my code I'm invoking `ContainerClient.ListBlobsHierarchy()` to iterate over a list of blobs in an Azure Blob container.
I'm using azure-sdk-for-go in combination with Azurite.

However, tests fail with error `get response from Azure API: no Host in request URL`.

I started troubleshooting and found out that `containerClient.ListBlobHierarchySegment()` on the next call of `NextPage()` is passing `test-fixtures-TestAzBlobClient_ListObjects-2021-11-26T114346.107911+0100/test.csv.1` into `runtime.NewRequest()` as value of parameter `endpoint`.

I'm wondering is it problem of Azurite which doesn't compose NextMarker properly or it's problem of ListBlobHierarchySegment() which doesn't respect `containerClient` configuration ?

Here is a screenshot of debugger demonstrating the process of troubleshooting:

![image](https://user-images.githubusercontent.com/380882/143572607-56762f55-ae5a-47f1-9dce-cd50c2c366ea.png)


My code:
```
                 perPageConv := int32(2)

		listOptions := &azblob.ContainerListBlobHierarchySegmentOptions{
			Prefix: &prefix,
			Maxresults: &perPageConv,
		}
		azPager := azClient.ListBlobsHierarchy("/", listOptions)

                for {
			hasNext := azPager.NextPage(ctx)
                         if azErr := azPager.Err(); azErr != nil { 
				fmt.Println("get response from Azure API: ", azErr.Error())
				break
                        }

			page := azPager.PageResponse()

			for i, commonPrefix := range page.Segment.BlobPrefixes {
                                 fmt.Println("Blob prefix ", commonPrefix.Name)
			}

			for i, keyObject := range page.Segment.BlobItems {
                                 fmt.Println("Blob ", keyObject.Name)
			}

			if !hasNext {
				break
			}

			pageNo++
		}
```

Thank you @adreed-msft for looking at this issue. I cannot test your PR as there are broken dependencies. Will test when your PR be merged.

```
$ go get github.com/Azure/azure-sdk-for-go/sdk/storage/azblob@adreed/fix-pagers
go: downloading github.com/Azure/azure-sdk-for-go/sdk/storage/azblob v0.1.1-0.20211201002034-a166c1381cd2
go: downloading github.com/Azure/azure-sdk-for-go v58.3.1-0.20211201002034-a166c1381cd2+incompatible
# github.com/Azure/azure-sdk-for-go/sdk/storage/azblob
../../../Library/Go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/storage/azblob@v0.1.1-0.20211201002034-a166c1381cd2/zc_append_blob_client.go:17:47: undefined: azcore.Credential
../../../Library/Go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/storage/azblob@v0.1.1-0.20211201002034-a166c1381cd2/zc_blob_client.go:17:9: undefined: azcore.Credential
../../../Library/Go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/storage/azblob@v0.1.1-0.20211201002034-a166c1381cd2/zc_blob_client.go:21:41: undefined: azcore.Credential
../../../Library/Go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/storage/azblob@v0.1.1-0.20211201002034-a166c1381cd2/zc_block_blob_client.go:31:46: undefined: azcore.Credential
../../../Library/Go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/storage/azblob@v0.1.1-0.20211201002034-a166c1381cd2/zc_connection_string.go:38:62: undefined: azcore.Credential
../../../Library/Go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/storage/azblob@v0.1.1-0.20211201002034-a166c1381cd2/zc_container_client.go:18:9: undefined: azcore.Credential
../../../Library/Go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/storage/azblob@v0.1.1-0.20211201002034-a166c1381cd2/zc_container_client.go:27:51: undefined: azcore.Credential
../../../Library/Go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/storage/azblob@v0.1.1-0.20211201002034-a166c1381cd2/zc_page_blob_client.go:23:45: undefined: azcore.Credential
../../../Library/Go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/storage/azblob@v0.1.1-0.20211201002034-a166c1381cd2/zc_service_client.go:30:9: undefined: azcore.Credential
../../../Library/Go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/storage/azblob@v0.1.1-0.20211201002034-a166c1381cd2/zc_service_client.go:40:47: undefined: azcore.Credential
../../../Library/Go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/storage/azblob@v0.1.1-0.20211201002034-a166c1381cd2/zc_service_client.go:40:47: too many errors
```

--- a/sdk/storage/azblob/zc_container_client.go
+++ b/sdk/storage/azblob/zc_container_client.go
@@ -7,6 +7,8 @@ import (
 	"context"
 	"time"
 
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+
 	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
 )
 
@@ -166,6 +168,13 @@ func (c ContainerClient) ListBlobsFlat(listOptions *ContainerListBlobFlatSegment
 		return pager
 	}
 
+	// override the advancer
+	pager.advancer = func(ctx context.Context, response ContainerListBlobFlatSegmentResponse) (*policy.Request, error) {
+		return c.client.listBlobFlatSegmentCreateRequest(ctx, &ContainerListBlobFlatSegmentOptions{
+			Marker: response.NextMarker,
+		})
+	}
+
 	// TODO: Come Here
 	//pager.err = func(response *azcore.Response) error {
 	//	return handleError(c.client.listBlobFlatSegmentHandleError(response))
@@ -189,8 +198,14 @@ func (c ContainerClient) ListBlobsHierarchy(delimiter string, listOptions *Conta
 		return pager
 	}
 
-	// TODO: Come here
-	//p := pager.(*listBlobsHierarchySegmentResponsePager)
+	// override the advancer
+	pager.advancer = func(ctx context.Context, response ContainerListBlobHierarchySegmentResponse) (*policy.Request, error) {
+		return c.client.listBlobHierarchySegmentCreateRequest(ctx, delimiter, &ContainerListBlobHierarchySegmentOptions{
+			Marker: response.NextMarker,
+		})
+	}
+
+	// todo: come here
 	//p.errorer = func(response *azcore.Response) error {
 	//	return handleError(c.client.listBlobHierarchySegmentHandleError(response))
 	//}

